src/Peachpie.Library/Hash.cs

```diff
@@ -3024,13 +3032,422 @@ public static string crypt(string str, string salt/* mandatory since 5.6 */)
+                }
+                else
+                {
+                    fieldhelper[0] = passwd[0];
+                    md5.Update(fieldhelper);
+                }
+            }
+
+            byte[] output = new byte[MD5MaxLength];
+            Array.Copy(magic, 0, output, 0, magic.Length);
+            Array.Copy(saltInBytes, 0, output, magic.Length, saltInBytes.Length);
+            output[magic.Length + saltInBytes.Length] = (byte)'$';
+
+            final = md5.Final();
+
+            for (int i = 0; i < 1000; i++)
+            {
+                md5Alt = new HashPhpResource.MD5();
+                md5Alt.Init();
+
+                if ((i & 1) != 0)
+                    md5Alt.Update(passwd);
+                else
+                    md5Alt.Update(final);
+
+                if ((i % 3) != 0)
+                    md5Alt.Update(saltInBytes);
+
+                if ((i % 7) != 0)
+                    md5Alt.Update(passwd);
+
+                if ((i & 1) != 0)
+                    md5Alt.Update(final);
+                else
+                    md5Alt.Update(passwd);
+
+                final = md5Alt.Final();
+            }
+
+            int length = magic.Length + saltInBytes.Length + 1;
+
+            int l;
+            l = (final[0] << 16) | (final[6] << 8) | final[12]; MD5MapASCII(output, length, l, 4); length += 4;
+            l = (final[1] << 16) | (final[7] << 8) | final[13]; MD5MapASCII(output, length, l, 4); length += 4;
+            l = (final[2] << 16) | (final[8] << 8) | final[14]; MD5MapASCII(output, length, l, 4); length += 4;
+            l = (final[3] << 16) | (final[9] << 8) | final[15]; MD5MapASCII(output, length, l, 4); length += 4;
+            l = (final[4] << 16) | (final[10] << 8) | final[5]; MD5MapASCII(output, length, l, 4); length += 4;
+            l = final[11]; MD5MapASCII(output, length, l, 2); length += 2;
+
+            return ConvertByteArrayToString(output.Take(length).ToArray());
+        }
+        #endregion
+
+        #region sha512/256 hash in crypt function
+        // This code in region was copied and modified from The PHP Interpreter (https://github.com/php/php-src/blob/master/ext/standard/crypt_sha256.c and crypt_sha512.c) 
+        private const string prefixRoundsSHA = "rounds=";
+        private const string prefixSaltSHA512 = "$6$";
+        private const string prefixSaltSHA256 = "$5$";
+        private const int saltMaxLength = 16;
+        private const int roundsDefault = 5000;
+        private const int roundsMin = 1000;
+        private const int roundsMax = 999999999;
+        // Table with characters for base64 transformation.
+        private const string b64t = "./0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+
+        /// <summary>
+        /// Part of cypt function in php. if sha 512 is false, sha 256 is used.
+        /// </summary>
+        private static string CryptSHA(Context ctx, string password, string salt, bool sha256)
```

Comment:
`Context ctx` is not needed

Reply:
Resolved https://github.com/peachpiecompiler/peachpie/pull/593/commits/6f1865f58f6166fda85c7352d98cab2d049e5ddf